File: Data Structures & Algorithms/maximum-product-subarray/submission-3.py
class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        #create an array to store results
        #Iterate through. When you hit Negative, branch off. When another negative is hit. collapse them together
        #If hit 0 before the split. just add the result to array and keep goin.
        #If hit 0 after the split then max of split is added. split is ended. and the pre-split max is added.
        res = max(nums)
        curMax, curMin = 1, 1

        for n in nums:
            temp = curMax * n
            curMax = max(curMax * n, curMin * n, n)
            curMin = min(temp, curMin*n, n)

            res = max(res, curMax)

        return res

File: Data Structures & Algorithms/jump-game-ii/submission-3.cpp
class Solution {
public:
    int jump(vector<int>& nums) {
        int res = 0;
        int q = nums.size();
        int r = 0; 
        int l = 0;
        int furthest;

        while(r < q-1)
        {
            furthest = 0;
            for(int i = l; i < r+1; i++)
            {
                furthest = max(furthest, i + nums[i]);
            }
            l = r+1;
            r = furthest;
            res++;
        }

        return res;
    }
};
